File: android/app/src/main/java/com/alandaitch/anden/data/catalog/StationCatalog.kt
package com.alandaitch.anden.data.catalog

import android.content.Context
import com.alandaitch.anden.AndenApp
import com.alandaitch.anden.data.model.Ramal
import com.alandaitch.anden.data.model.Station
import com.alandaitch.anden.data.model.TrainLine
import com.alandaitch.anden.util.Geo
import com.alandaitch.anden.util.GeoPoint
import com.alandaitch.anden.util.Text
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// Catálogo embebido de estaciones y ramales. Carga estaciones.json + lineas.json de assets.
class StationCatalog(context: Context = AndenApp.appContext) {

    val all: List<Station>
    val lines: List<TrainLine> = TrainLine.all

    private val stationsById: Map<Int, Station>
    private val ramalsById: Map<Int, Ramal>

    @Serializable
    private data class LineFile(val id: Int, val nombre: String, val ramales: List<Ramal> = emptyList())

    init {
        val json = Json { ignoreUnknownKeys = true }
        val stations: List<Station> = try {
            val text = context.assets.open("estaciones.json").bufferedReader().use { it.readText() }
            json.decodeFromString(text)
        } catch (_: Exception) {
            emptyList()
        }
        all = stations
        stationsById = stations.associateBy { it.id }

        val ramals = mutableMapOf<Int, Ramal>()
        try {
            val text = context.assets.open("lineas.json").bufferedReader().use { it.readText() }
            val files: List<LineFile> = json.decodeFromString(text)
            for (f in files) for (r in f.ramales) ramals[r.id] = r
        } catch (_: Exception) {
        }
        ramalsById = ramals
    }

    fun station(id: Int): Station? = stationsById[id]

    fun ramal(id: Int): Ramal? = ramalsById[id]

    // Substring sin acentos. Prioriza enRamalPublico y tieneArribosHoy.
    fun search(query: String): List<Station> {
        val q = normalize(query)
        if (q.isEmpty()) return emptyList()
        return all
            .filter { it.visibleEnApp && normalize(it.nombre).contains(q) }
            .sortedWith(compareByDescending<Station> {
                (if (it.enRamalPublico) 2 else 0) + (if (it.tieneArribosHoy) 1 else 0)
            }.thenBy { it.nombre.lowercase() })
    }

    // Estaciones más cercanas con servicio real, ordenadas por distancia (metros).
    fun nearest(to: GeoPoint, limit: Int = 5): List<Pair<Station, Double>> {
        return all
            .filter { it.enRamalPublico }
            .map { it to Geo.distanceMeters(it.coordinate, to) }
            .sortedBy { it.second }
            .take(limit)
    }

    companion object {
        val shared: StationCatalog by lazy { StationCatalog() }

        fun normalize(s: String): String = Text.normalize(s)
    }
}

File: Shared/Catalog/StationCatalog.swift
import Foundation
import Observation
import CoreLocation

// Catálogo embebido de estaciones y ramales. Carga estaciones.json + lineas.json del bundle.
@Observable
final class StationCatalog {
    static let shared = StationCatalog()

    private(set) var all: [Station] = []
    private(set) var lines: [TrainLine] = TrainLine.all

    private var stationsById: [Int: Station] = [:]
    private var ramalsById: [Int: Ramal] = [:]

    init(bundle: Bundle = .main) {
        load(bundle: bundle)
    }

    private struct LineFile: Decodable {
        let id: Int
        let nombre: String
        let ramales: [Ramal]
    }

    private func load(bundle: Bundle) {
        let dec = JSONDecoder()
        if let url = bundle.url(forResource: "estaciones", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let stations = try? dec.decode([Station].self, from: data) {
            all = stations
            stationsById = Dictionary(stations.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        }
        if let url = bundle.url(forResource: "lineas", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let files = try? dec.decode([LineFile].self, from: data) {
            for f in files {
                for r in f.ramales { ramalsById[r.id] = r }
            }
        }
    }

    func station(id: Int) -> Station? { stationsById[id] }

    func ramal(id: Int) -> Ramal? { ramalsById[id] }

    // Substring sin acentos. Prioriza enRamalPublico y tieneArribosHoy.
    func search(_ query: String) -> [Station] {
        let q = Self.normalize(query)
        guard !q.isEmpty else { return [] }
        return all
            .filter { $0.visibleEnApp && Self.normalize($0.nombre).contains(q) }
            .sorted { a, b in
                let ra = (a.enRamalPublico ? 2 : 0) + (a.tieneArribosHoy ? 1 : 0)
                let rb = (b.enRamalPublico ? 2 : 0) + (b.tieneArribosHoy ? 1 : 0)
                if ra != rb { return ra > rb }
                return a.nombre.localizedCaseInsensitiveCompare(b.nombre) == .orderedAscending
            }
    }

    // Estaciones más cercanas con servicio real, ordenadas por distancia.
    func nearest(to coordinate: CLLocationCoordinate2D, limit: Int = 5) -> [(Station, CLLocationDistance)] {
        let origin = CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)
        return all
            .filter { $0.enRamalPublico }
            .map { ($0, CLLocation(latitude: $0.lat, longitude: $0.lng).distance(from: origin)) }
            .sorted { $0.1 < $1.1 }
            .prefix(limit)
            .map { $0 }
    }

    static func normalize(_ s: String) -> String {
        s.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "es"))
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
